sa/sa.go

```diff
@@ -2070,3 +2070,40 @@ func (ssa *SQLStorageAuthority) KeyBlocked(ctx context.Context, req *sapb.KeyBlo
 
 	return &sapb.Exists{Exists: true}, nil
 }
+
+// IncidentsForSerial returns a list of active incidents for `req.Serial`.
+func (ssa *SQLStorageAuthority) IncidentsForSerial(ctx context.Context, req *sapb.Serial) ([]sapb.Incident, error) {
+	if req == nil {
+		return nil, errIncompleteRequest
+	}
+
+	var activeIncidents []incidentModel
+	_, err := ssa.dbMap.Select(&activeIncidents, `SELECT * FROM incidents WHERE enabled = 1`)
+	if err != nil {
+		if db.IsNoRows(err) {
+			return nil, berrors.NotFoundError("no active incidents found")
+		}
+		return nil, err
+	}
+
+	var incidentsForSerial []sapb.Incident
+	for _, i := range activeIncidents {
+		var count int
+		err := ssa.dbMap.SelectOne(&count, fmt.Sprintf("SELECT COUNT(*) FROM %s WHERE serial = %s",
```

Comment:
```suggestion
		err := ssa.dbMap.SelectOne(&count, fmt.Sprintf("SELECT COUNT(*) FROM %s WHERE serial = ?",
```

In general we want to always use `?` interpolation (which relies on MariaDB's prepared statements, and is nice and safely escaped). We want to never use `%s` interpolation because it bypasses our defenses against SQL injection. However, there are some unavoidable exceptions.

It's okay to interpolate the table name using `%s` here because (a) it comes from an internal, trusted source (the incidents table), and (b) there's not other way to make this query generic across tables.

However, it's not okay to interpolate the serial number using `%s`, because it's potentially a parameter all the way from the user (or from an attacker partway into the network). Instead we should use `?` for the serial.